### ISSUE TITLE
What is leading to a transactions rejected as "stale"

### ISSUE THREAD

**Issue by Tbaut** [polkadot-api/polkadot-api#814]
I'm porting this testing tool to papi: https://github.com/ChainSafe/cypress-polkadot-wallet
To test this plugin, I am creating and signing 2 transactions consecutively, those are 2 txs in 2 different blocks.
I do not wait for finalization, but my `signSubmitAndWatch` uses the `{ at: 'best' }` here:
https://github.com/ChainSafe/cypress-polkadot-wallet/blob/7fc2b970520bbdaba97e2094a0b77336d19a60d1/packages/example/src/main.ts#L142-L143

to help you understand what I do, the test consist in:
- the page loads
- the Dapp connects to this emulated extension
- we verify that there are accounts connected
- we sign a tx and broadcast it
- we wait for some events from the `txBestBlocksState`

Once this is done, and the expected event is found, we do it again, in a new test, loading the page again and starting from scratch.
Independently both tests work without issue. One is supposed to work well (ExtrinsicSuccess) and the other to fail (not enough funds).

Now, when run one after the other, the second fails with "stale". This all happens in <30s, and from my point of view, it looks like the `nonce` is outdated the second time the test runs, despite the `best`.

Could it be that `at: best` isn't doing what I'd expect it to do. How can I check?

**Comment by josepot** [comment]
> What is leading to a transactions rejected as "stale"

A nonce that's behind the knowledge of a full-node.

> I do not wait for finalization, but my `signSubmitAndWatch` uses the `{ at: 'best' }` here:

In this case, you must handle the nonce yourself. Otherwise, both transactions will have the same nonce.

There is an interesting discussion about why this is the case [in here](https://github.com/paritytech/json-rpc-interface-spec/issues/156).

It's annoying, but the solution would be to query the nonce yourself before creating the transactions, and then pass the correct nonce to each transaction. Otherwise, PAPI will take the nonce from the state on-chain, which has not being altered yet.

The real solution to this problem is to have the "signer"/extension handle the nonce. The reason being that the signer knows about all the transactions that are being created by all different DApps. So, that's what should really keep track of the nonce and handle the broadcasting once it's safe, etc. This is one of the main reasons why we need a proper "Polkadot Provider" interface... I think that one day we will have it... A man can dream 🙂

**Comment by Tbaut** [comment]
> In this case, you must handle the nonce yourself. Otherwise, both transactions will have the same nonce.

I thought having the `best` in `signSubmitAndWatch` would use the best block for the nounce. Am I wrong?

Just to be clear my tx are submitted in sequence, at least 1 block apart
- tx 1 submission
- wait for `txBestBlocksState`
- tx 2 submission (in a new block, that did not reorg)

**Comment by josepot** [comment]
> I thought having the `best` in `signSubmitAndWatch` would use the best block for the nounce. Am I wrong?

you are not wrong. No.

However, when you create 2 transaction in a row, the second transaction is not yet part of the best-block chain... So, then both of them get the same nonce.

> Just to be clear my tx are submitted in sequence, at least 1 block apart
> 
> * tx 1 submission
> * wait for `txBestBlocksState`
> * tx 2 submission (in a new block, that did not reorg)

We have E2E tests that test this flow. So, I'm quite certain that this works. Are you sure that you are waiting for the first transaction to be included in the best-block chain? If that's the case, then my recommendation would be to create the transaction on top of that block (ie: don't pass best on the second transaction, just pass the block hash of the block where the first transaction is included).

**Comment by Tbaut** [comment]
> Are you sure that you are waiting for the first transaction to be included in the best-block chain?

Yes I'm certain, my test is verifying that the UI shows elements (a multisig tx) that can only result in a tx being in best block and the state of the blockchain up to date before we more to the next test. 

I have tried to reproduce, and am facing another issue, `Bad proof` which I don't understand. it's here: https://github.com/Tbaut/papi-node/blob/2-txs/src/main.ts

the test is quite straight-forward, and result in having the 2nd tx not broadcasted due to "BadProof"

```ts
  const firstTx = api.tx.System.remark({
    remark: Binary.fromText('hello world 1'),
  });

  // create the 1st signed extrinsic
  const signedTx = await firstTx.sign(aliceSigner, { at: 'best' });

  await client.submitAndWatch(signedTx).subscribe({
    next: async (event) => {
      if (event.type === 'txBestBlocksState' && event.found === true) {
        console.log('1 in block');

        // create the 1nd signed extrinsic
        const secondTx = api.tx.System.remark({
          remark: Binary.fromText('hello world 2'),
        });

        const secondSignedTx = await secondTx.sign(aliceSigner, { at: 'best' });

        client.submitAndWatch(secondSignedTx).subscribe({
          next: (e) => console.log(2, JSONprint(e)),
          error: (e) => console.log(2, JSONprint(e)),
        });
      }
    },
    error: (e) => console.log(1, JSONprint(e)),
  });
```

**Comment by josepot** [comment]
I'm reopening this, b/c according to @Tbaut  there seems to be a race condition in here. Also, I just noticed that the `txBestBlocksState` event doesn't include the block info when `found` is `true` 😞

**Comment by Tbaut** [comment]
Fixed by using `signSubmitAndWatch` instead. I checked and it worked with 1.6.2 too.. so I can't take any credit for reporting the bug you fixed in 1.6.4 :thinking:
